compat/openssl.h

```diff
@@ -18,6 +18,7 @@
 
 #ifndef OPENSSL_COMPAT_H
 #define OPENSSL_COMPAT_H
+#if !IN_TESTHEADERS
```

Comment:
Why do we need this? Are there any viable alternatives to sprinkling non-test code with these test exclusions? If there are none, what non-test code will qualify for an exclusion like that?

Reply:
One of the problems is:
We're not supposed to include this header unless USE_OPENSSL is defined; this should be include-guarded in the including file.
But when testing the headers, we unconditionally include this.
There's two ways around that: either special-casing in Makefile.am, or special-casing in the file itself. This condition is an implementation of the latter approach

Reply:
> There's two ways around that: either special-casing in Makefile.am, or special-casing in the file itself.

Agreed. However, that "special-casing in the file itself" does not have to use IN_TESTHEADERS. It can use USE_OPENSSL instead (with trivial refactoring to remove the current `#error`). Thus, there are at least three options here. The fourth solution is to move this header file into src/ssl/, which is already compiled conditionally.

I believe the two best solutions here are moving the file and special-casing in Makefile.am because "make check" should not be testing headers that "make" is not using (roughly speaking). The latter principle may be useful elsewhere in this PR as well.

Reply:
even after going the SOURCES route, compat/openssl.h is unconditionally referenced as part of libcompatsquid_la_SOURCES

Reply:
I am resolving this change request because the problematic change is now gone. The current alternative solution in this PR (i.e. what I call "double guarding") is also problematic IMO, but that problem is discussed in [another change request](https://github.com/squid-cache/squid/pull/1463#discussion_r1333133479).